### PR TITLE
Try connecting in tests

### PR DESCRIPTION
**Pull request by eagr**
The rough idea of what we talked about. :)) @liamaharon

### CODE REVIEW

**Comment by liamaharon** [comment]
Looking good. 

It would be nice if we could specify a block number we'd like it to run to before proceeding as well.

**Comment by eagr** [comment]
Could you elaborate a bit on "a block number we'd like it to run to"?

**Comment by liamaharon** [comment]
Poll the current block of the spun up chain and yield after a block has been mined.

**Comment by eagr** [comment]
You mean something like this?

```rs
async fn block_hash(from: u64, to: u64, url: &str, timeout: u64) -> Result<Vec<Hash>, String>;
```

File: core/tests/create_snapshot.rs [hunk 1]
```diff
@@ -56,7 +54,43 @@ async fn create_snapshot_works() {
...
+                tokio::time::sleep(Duration::from_secs(interval)).await;
+                interval = (interval / 3 * 2).max(8);
+            };
+
+            let result = ChainApi::<(), Hash, Header, ()>::block_hash(
+                &rpc,
+                Some(ListOrValue::Value(NumberOrHex::Number(block_number))),
+            )
+            .await
+            .map_err(|_| "Couldn't get block hash".to_string())?;
```
Comment (liamaharon) [inline core/tests/create_snapshot.rs:85]: What I mean is by the time you get there, the block may not have been mined yet so probably needs to continue waiting *after* it has connected until confirmation that the target block is mined

**Comment by eagr** [comment]
By grouping the integration tests, on my machine, it could save ~70s and ~1GB disk space. In case it's not obvious from the report, this essential reduces the amount of build targets, therefore reduces the overall amount of work (linking) and generated artifacts.

Original
image

Grouped
image

**Comment by eagr** [comment]
```
Thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: 
GlobalLoggerError(SetLoggerError(SetLoggerError(())))',
/Users/eagr/.cargo/git/checkouts/polkadot-sdk-cff69157b985ed76/1d5d4a4/substrate/test-utils/cli/src/lib.rs:66
```

I suspect this is due to we're running the tests in parallel now, spinning up multiple dev nodes at the same time. Any ideas?

**Comment by eagr** [comment]
This would solve the issue, and the tests also run faster this way. But I'm not sure whether this is what you want. The logs are all mixed up as the tests run in parallel now.

**Comment by eagr** [comment]
Currently, the tests run sequentially in separate processes. There are no problems because by the time the next test runs, the dev node has exited with the previous test process.

After grouping the tests run in one process concurrently (different threads). The problem originates from setting the logger multiple times in a process, which is disallowed by the tracing crate.

https://docs.rs/tracing-log/latest/tracing_log/log_tracer/struct.SetLoggerError.html

Making the tests run sequentially (either with lock or `--test-threads=1`) wouldn't work, unless there is a way to force the dev node to exit at the end of a test.

TL;DR Grouping the tests has the benefits of faster compilation (-1min) and execution (-3min), but if running tests in parallel on the same node isn't what you want, let's revert the changes.

**Comment by rockbmb** [comment]
@eagr thanks for this contribution!

I'm closing this PR as it's been open for a while.
Future contributions are always welcome.
